### ISSUE TITLE
use of 'pytest' in setup.cfg collides with 'pytest' distutils command

### ISSUE THREAD

**Issue by pytestbot** [pytest-dev/pytest#567]
Originally reported by: **Jason R. Coombs (BitBucket: [jaraco](http://bitbucket.org/jaraco), GitHub: [jaraco](http://github.com/jaraco))**

----------------------------------------

In [pytest-runner 2](/jaraco/pytest-runner/issue/2), I adapted pytest-runner to support the 'pytest' distutils command. However, because the pytest project itself uses the ``[pytest]`` section of setup.cfg, that conflicts with the same section for the 'pytest' distutils command.

Using the 'pytest' distutils command in a project which specifies, for example, 'norecursedirs', the runner will fail because the pytest-runner distutils command doesn't recognize that parameter (or any other ini options). An example error message is:

```
error: error in setup.cfg: command 'PyTest' has no such option 'norecursedirs'
```

I see a few options here:

1. Just use the 'ptr' command name.
2. Have pytest runner distutils command disregard all unrecognized options.
3. Have pytest runner distutils command implement and ignore all of the options that might be presented in setup.cfg. It would need to do this before pytest is present (because args are processed before dependencies such as pytest are downloaded/imported).
4. Have pytest drop support for setup.cfg.
5. Require that users not use setup.cfg for pytest ini options (only support pytest.ini and tox.ini for pytest options when running under distutils).

I suspect there are other options, too. I'm not particularly happy with any of those options, but I'm leaning toward (2).

I've filed the ticket here with pytest for two reasons:

- I want the implementation to be as acceptable as possible for the pytest project to endorse it as a viable integration mechanism.
- I believe the use of the [pytest] section in the setup.cfg by the pytest library is a violation of the [explicit expectation](https://docs.python.org/2/distutils/configfile.html) that those sections are meant for distutils commands.

@hpk42 What is your reaction? What would you suggest?

----------------------------------------
- Bitbucket: https://bitbucket.org/pytest-dev/pytest/issue/567

**Comment by RonnyPfannschmidt** [comment]
We should deprecate and drop usage in setup.cfg

**Comment by nicoddemus** [comment]
Related to #1311

**Comment by ches** [comment]
All said up to this point in *this* thread, I probably would have voted for option 1: don't use `pytest` as `pytest-runner`'s distutils command name. A project previously using `[pytest]` in `setup.cfg` without issue has a broken experience when trying to add `pytest-runner` to the project and running `setup.py test`—from that perspective, the user considers `pytest-runner` as the broken piece, and the responsibility lies with `pytest-runner` to fix that at the peril of losing those users. It's not ideal, but it's a straightforward solution, and I wager nearly everyone immediately aliases to `test` anyway.

If there are other problems, though (#1311), then I still like @RonnyPfannschmidt's suggestion of changing to a prefix like `[tool:pytest]` if that is workable. My preference for this probably aligns with @hpk42's original motive mentioned above—proliferation of too many config and metadata files for tools. Others support `setup.cfg` as well (flake8, nose I believe), so it was nice to consolidate some of the root directory cruft in new projects which not only feels messy but IMO creates confusion for newcomers less familiar with the Python ecosystem. This is of course a subjective issue beyond the scope of this matter, but using `setup.cfg` *seemed* an innocent solution.

If this is deemed an abuse of distutils' turf, I can see the argument to remove `setup.cfg` support altogether, but right or wrong it has become pretty widespread at this point. Having originally opened this issue and being a PyPA contributor, maybe @jaraco has more opinion about this now.

@foxx's problem report in #1311 is a little unclear to me: it appears he was *using* `pytest-runner`, which possibly conflates things. Not sure if the `addopts` issue was when invoking through the runner exclusively or not.

**Comment by ches** [comment]
FWIW, Coverage [is another that supports `setup.cfg`][1], using a prefix pattern.

[1]: http://coverage.readthedocs.io/en/coverage-4.1/config.html

**Comment by RonnyPfannschmidt** [comment]
@ches thanks for the reference, but as you can see, the "sane" tools do use a prefix while pytest does not

**Comment by jaraco** [comment]
My opinion remains basically the same, though it's crystallized even more now. The namespace for non-prefixed names in setup.cfg is practically if not explicitly reserved for Distutils command names, so pytest should drop support for it or use a prefixed name.

I would be happy to switch pytest-runner back to using the 'ptr' name (it's currently still supported), but only if there's a consensus to do so, and I still feel like that's a less suitable solution than the alternative.

**Comment by ssbarnea** [comment]
Hi guys! I ended up to this page shortly after trying to implement the good practices from http://docs.pytest.org/en/latest/goodpractices.html

What should I do to pass over this issue? ...and be able to run tests using the standard  `python setup.py test`. Do we have any proof-of-concept examples?

I am using pytest and partially tox (tox when doing matrix builds locally and just pytest when on travis as it has its own matrix setup). Thanks.

**Comment by ches** [comment]
@ssbarnea You should be fine following the advice of the documentation to use pytest-runner, just note that you can't use a `[pytest]` section in `setup.cfg` unless the options you're setting there are explicitly supported by pytest-runner. You will instead want to use `pytest.ini` or `tox.ini` for your general pytest configuration.

**Comment by The-Compiler** [comment]
This one seems like an easy, backwards-incompatible change for 3.0 - change `[pytest]` to `[tool:pytest]` in `setup.cfg` and warn/deprecate about `[pytest]`?

**Comment by jaraco** [comment]
@The-Compiler And introduce forward-compatibilty for `[tool:pytest]` ASAP.

**Comment by The-Compiler** [comment]
@jaraco Yeah - we want to release 3.0 ASAP, I was looking at which issues really should go into 3.0 (because otherwise they'd need to be postponed to 4.0).

**Comment by hackebrot** [comment]
Anyone capacities to pick this one up?

**Comment by nicoddemus** [comment]
@hackebrot I have to remove invocation fixtures first and was going to get this if nobody volunteers until then, but if you have the time and want to tackle this it would be great.

Just to make sure I understand, the plan is to add support for `[tool:pytest]` in `setup.cfg` and make that the official and preferred way to specify pytest options in a `setup.cfg` file. At the same time, emit a deprecation warning if it encounters a plain `[pytest]` section in a `setup.cfg` file. Nothing should be done regarding the current way to specify options in `pytest.ini` or `tox.ini` files.

Is my understanding correct?

**Comment by hackebrot** [comment]
I am busy with #1795 atm, but I'd like to see this being worked on. 😄 

Can't wait to use pytest 3.0. 🚀

**Comment by nicoddemus** [comment]
Resolved in #1813
